lib/ssl-helper.js

```diff
@@ -2,16 +2,14 @@ import tls from 'tls';
 import net from 'net';
 import B from 'bluebird';
 
-const TLS_VERSION = 'TLSv1_method';
 const HANDSHAKE_TIMEOUT_MS = 10000;
 
 function upgradeToSSL (socket, key, cert) {
   return new tls.TLSSocket(socket, {
     rejectUnauthorized: false,
     secureContext: tls.createSecureContext({
       key,
-      cert,
-      secureProtocol: TLS_VERSION
+      cert
```

Comment:
I assume we still need to set `minVersion` to `TLSv1` as it is equal to `TLSv1.2` by default

Reply:
I can confirm that neither "TLSv1.1" nor "TLSv1" will work on iOS 14.5.1

quick test script:
```
import {utilities } from 'appium-ios-device'

const udid = await utilities.getConnectedDevices();
console.log(udid[0])
const lockdown = await utilities.startLockdownSession(udid[0])

const syslog_relay = await lockdown.startService('com.apple.syslog_relay');
console.log(syslog_relay)
```
result:
```
[****/node_modules/appium-ios-device/lib/plist-service/index.js:67]()
        reject(new Error(`Failed to receive any data within the timeout: ${timeout}`));
               ^
node_modules/source-map-support/source-map-support.js:496
Error: Failed to receive any data within the timeout: 5000
    at Timeout._onTimeout ([****/node_modules/appium-ios-device/lib/plist-service/index.js:67:16]())
    at listOnTimeout (node:internal[/timers:559:17]())
    at processTimers (node:internal[/timers:502:7]())
node_modules/source-map-support/source-map-support.js:499
Process exited with code 1
```

Reply:
Interesting... Could you share your env?
I haven't tested with ios 14, but my iPhone SE (15.4), SE 2 (15.3.1) and iPad mini (15.4) have no issue with current implementation via macOS 12.2.1. I wonder which factor could be related to this timeout issue.

Reply:
Did you install `node` using `brew`? The `brew` version of `node` doesn't seem to respect `minVersion: 'TLSv1'`, which I suspect has something to do with node's openssl compile options. The bug can be reproduced with any version of node installed via `nvm` on macOS 11.6 ~ 12.2.1 (intel and M1).

Reply:
AFAIK minVersion is just a hint for the protocol. It does not require the TLSSocket to use this version, but just limits the lowest number:

> minVersion [<string>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) Optionally set the minimum TLS version to allow. One of 'TLSv1.3', 'TLSv1.2', 'TLSv1.1', or 'TLSv1'. Cannot be specified along with the secureProtocol option; use one or the other. Avoid setting to less than TLSv1.2, but it may be required for interoperability. Default: [tls.DEFAULT_MIN_VERSION](https://nodejs.org/api/tls.html#tlsdefault_min_version).

My point is that since it's set to `TLSv1.2` by default then the compatibility with older iOS devices might be broken.

Reply:
I assume a similar effect would be achieved if `secureProtocol` is set to `TLS_method`

Reply:
It seems that it has nothing to do with iOS, but is caused by the difference between openssl 3.0 and openssl 1.1.1, may be a openssl 3.0 bug?

`nvm` version of `node v17.6.0`, timeout error:
```
$ node        
Welcome to Node.js v17.6.0.
Type ".help" for more information.
> process.versions
{
  node: '17.6.0',
  v8: '9.6.180.15-node.13',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '102',
  nghttp2: '1.45.1',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '3.0.1+quic',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}
```
`nvm` version of `node v16.14.0`, no error:
```
$ node
Welcome to Node.js v16.14.0.
Type ".help" for more information.
> process.versions
{
  node: '16.14.0',
  v8: '9.4.146.24-node.20',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.45.1',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '1.1.1m+quic',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}
```
`brew` version of `node v17.7.1`, no error:
```
$ /usr/local/opt/node@17/bin/node
Welcome to Node.js v17.7.1.
Type ".help" for more information.
> process.versions
{
  node: '17.7.1',
  v8: '9.6.180.15-node.15',
  uv: '1.44.1',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '102',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '1.1.1m',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0'
}
```
`brew` version of `node v16.14.0`, no error:
```
$ /usr/local/opt/node@16/bin/node
Welcome to Node.js v16.14.0.
Type ".help" for more information.
> process.versions
{
  node: '16.14.0',
  v8: '9.4.146.24-node.20',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.46.0',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '1.1.1m',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0'
}
```

Reply:
could be...
Try to report it to OpenSSL issues tracker and see what the response would be.
We can also try to workaround it by checking the actual OpenSSL version number. This would be ugly, but something is better than nothing

Reply:
Oh... good to find...
I confirmed my env is 1.1.1. I use nvm and nodebrew to handle node versions. Never used view brew to install nodejs.

Reply:
may be related to this
https://github.com/openssl/openssl/blob/openssl-3.0.1/doc/man7/migration_guide.pod#user-content-tls-changes
> The security strength of SHA1 and MD5 based signatures in TLS has been reduced.
>
> This results in SSL 3, TLS 1.0, TLS 1.1 and DTLS 1.0 no longer working at the default security level of 1 and instead requires security level 0. The security level can be changed either using the cipher string with @SECLEVEL, or calling [SSL_CTX_set_security_level(3)](http://man.he.net/man3/SSL_CTX_set_security_level). This also means that where the signature algorithms extension is missing from a ClientHello then the handshake will fail in TLS 1.2 at security level 1. This is because, although this extension is optional, failing to provide one means that OpenSSL will fallback to a default set of signature algorithms. This default set requires the availability of SHA1.

Reply:
By decompiling the `nvm` version of `node v17.6.0`, I can confirm that OPENSSL_TLS_SECURITY_LEVEL is `1`. and by binary patching to `0`, the timeout bug went away. So this is an openssl 3.0 feature and not a bug...
https://github.com/openssl/openssl/blob/openssl-3.0.1/ssl/ssl_cert.c#L63

<img width="1080" alt="image" src="https://user-images.githubusercontent.com/3118816/158027697-4d5462f3-60eb-44fe-a03c-ecbfd3549f63.png">

Reply:
thanks for the detailed investigation. I appreciate the level of details. Could you propose any simpler workaround to this rather than the binary patching?

Reply:
What I've found so far is that this var might be also picked up from the command line, like: `export OPENSSL_TLS_SECURITY_LEVEL=1`. Could you please confirm that @douniwan5788 ?

Reply:
`OPENSSL_TLS_SECURITY_LEVEL=0` env doesn't work, I tried `OPENSSL_TLS_SECURITY_LEVEL=0 node test.mjs` and it still timeout, I think it's a compile-time option, used when compiling node.

Apple supports TLS 1.2 since [iOS 5](https://developer.apple.com/library/archive/technotes/tn2287/_index.html), and [automatically applied to apps that are compiled for iOS 9 or later and macOS 10.11 or later](https://support.apple.com/guide/security/tls-security-sec100a75d12/web#secfb58b3350). I don't think any device that only supports tls 1.0/1.1 will work normally these days

Reply:
cool, then then defaults should work. Could you also please sign the CLA before we merge the patch?

Reply:
I found iOS 13.4.1 device. It worked with minVersion: 'TLSv1.2', so at least tls v1.2 should be enough for our usage.
Just a note.